tests/optim/test_scheduler.py

```diff
@@ -409,24 +412,62 @@ def test_warmup_schedulers_fail_fast(
         'dur': f'{warmup_duration_pct}dur',
     }
 
+    scheduler_max_unit_to_str = {
+        'tok': f'{max_tokens}tok',
+        'sp': f'{max_samples}sp',
+        'ba': f'{max_batches}ba',
+        'ep': f'{max_epochs}ep',
+        'dur': f'{scheduler_max_pct}dur',
+    }
+
     max_duration_str = max_duration_unit_to_str[max_duration_unit]
     warmup_duration_str = warmup_duration_unit_to_str[warmup_duration_unit]
+    scheduler_max_str = scheduler_max_unit_to_str[scheduler_max_unit]
     num_steps = total_batches
 
     if scheduler_class == MultiStepWithWarmupScheduler:
         scheduler = scheduler_class(milestones=['60ba'], t_warmup=warmup_duration_str)  # type: ignore
     else:
-        scheduler = scheduler_class(t_warmup=warmup_duration_str)  # type: ignore
+        scheduler = scheduler_class(t_warmup=warmup_duration_str, t_max=scheduler_max_str)  # type: ignore
 
     state = dummy_schedulers_state
     state.max_duration = Time.from_timestring(max_duration_str)
     state.timestamp = Timestamp()
     state.set_dataloader([None] * batches_per_epoch, 'train')
 
+    effective_scheduler_max_unit = scheduler_max_unit if scheduler_max_unit != 'dur' else max_duration_unit
+    effective_warmup_duration_unit = warmup_duration_unit if warmup_duration_unit != 'dur' else max_duration_unit
+    max_duration_no_epoch = state.max_duration
+    if max_duration_unit == 'ep':
+        max_duration_no_epoch = Time.from_timestring(max_duration_unit_to_str['ba'])
+    error_context = contextlib.nullcontext()
+    if (
+        hasattr(scheduler, 't_max') and
+        not _units_comparable(effective_scheduler_max_unit, effective_warmup_duration_unit)
+    ):
+        error_context = pytest.raises(ValueError, match='Cannot use warmup scheduler')
+    elif (
+        hasattr(scheduler, 't_max') and
+        _units_comparable(effective_scheduler_max_unit, effective_warmup_duration_unit) and
+        _units_comparable(effective_scheduler_max_unit, max_duration_unit) and
+        _convert_time(scheduler_max_str, state) < max_duration_no_epoch
+    ):
+        error_context = pytest.raises(ValueError, match='must be greater than or equal to max_duration')
+
     with error_context:
         for _ in range(num_steps):
             _ = scheduler(state)
             state.timestamp = state.timestamp.to_next_batch(
                 samples=samples_per_batch,
                 tokens=tokens_per_sample * samples_per_batch,
             )
+
+
+def _units_comparable(unit1: str, unit2: str) -> bool:
```

Comment:
can you use the `TimeUnit` enums here instead
https://github.com/mosaicml/composer/blob/2f11230a17bd35eb4f6a5784a18e43c648a56b2f/composer/core/time.py#L32

Reply:
I'll leave this to a separate PR since pytest params should also use TimeUnit enums if this change is made